src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/planning/SpaceBookingDayPlanner.kt

```diff
@@ -0,0 +1,147 @@
+package uk.gov.justice.digital.hmpps.approvedpremisesapi.service.cas1.planning
+
+import org.springframework.stereotype.Service
+
+@Service
+class SpaceBookingDayPlanner {
+  /**
+   * Given a list of available beds and required bookings for a given day, will attempt to optimally
+   * assign bookings to beds to determine a premise's capacity on a given day.
+   *
+   * We do not currently intend to present the 'booking to room', instead we are only using this planning
+   * to determine the number of free spaces after bookings are planned.
+   *
+   * If we were to present 'booking to room' planning, we may wish to discuss further optimisation
+   * with premises e.g. would they want to plan bookings into the fullest multiple occupancy room possible,
+   * Or is the goal to minimise shared rooms? What's more important, minimising characteristic surplus, or
+   * minimising single bookings into room with the least other beds?
+   *
+   * Planning goals:
+   *
+   * <ol>
+   *   <li>Priority is given to bookings with constraints, those with the most constraints being planned first</li>
+   *   <li>Bookings are placed into rooms with the least surplus characteristics</li>
+   *   <li>Single occupancy bookings are never placed in rooms with other bookings</li>
+   *   <li>If there are multiple competing rooms (after above goals are satisfied), single occupancy bookings will
+   *   be placed into rooms with the least beds (ensuring minimum 'wastage' in these scenarios)</li>
+   * </ol>
+   *
+   * @param beds All beds available on the day being planned (i.e. won't include out of service beds or beds inactive according to start/end date)
+   * @param bookings All required bookings. The code assumes that the premise level characteristics have already been met
+   */
+  fun plan(
+    beds: Set<Bed>,
```

Comment:
Shall we add in some error handling for poor data? Do you think there could be a circumstance in which there are two beds with the same reference, which are the same bed but have been incorrectly inputted?

Reply:
Whilst `bed.reference` being duplicated would produce confusing 'dev only' output from the planner (e.g. the text output we use to verify behaviour), it's actually `bed.id` that would cause an issue if duplicated.  I'm inclined not to add error handling as we trust the source (our own code), and if bed names/labels were duplicated it should be very obvious on the UI and be dealt with quickly. With that said, there's no harm in adding error handling if this is something you'd like adding